docs/DesignDocs/OpteeIntegration.md

```diff
@@ -0,0 +1,321 @@
+  uint32_t algo;
+  uint16_t hash_size;
+  uint16_t sig_size;
+
+  /*
+   * Dynamically-sized arrays:
+   *
+   * uint8_t hash[hash_size];
+   * uint8_t sig[sig_size];
+   */
+};
+
+struct shdr_bootstrap_ta {
+  uint8_t uuid[sizeof(TEE_UUID)];
+  uint32_t ta_version;
+};
+
+struct shdr_encrypted_ta {
+  uint32_t enc_algo;
+  uint32_t flags;
+  uint16_t iv_size;
+  uint16_t tag_size;
+
+  /*
+   * Dynamically-sized arrays:
+   *
+   * uint8_t iv[iv_size];
+   * uint8_t tag[tag_size];
+   */
+};
+```
+
+Encrypting TAs is optional.
+For an encrypted TA, the hash is computed as follows:
+```
+H = SHA_256(shdr | shdr_bootstrap_ta | shdr_encrypted_ta | standard ELF)
+```
+
+For a cleartext TA, the hash is computed as follows:
+```
+H = SHA_256(shdr | shdr_bootstrap_ta | standard ELF)
+```
+
+The signature is computed as follows:
+```
+S = RSA_SSA_PKCS1_PSS_MGF1_SHA256(H)
+```
+
+The ciphertext is computed as follows:
+```
+C = AES_GCM(standard ELF)
+```
+
+For an encrypted TA, the final TA image is composed as follows:
+```
+TA = (shdr | H | S | shdr_bootstrap_ta | shdr_encrypted_ta | C)
+```
+
+For a cleartext TA, the final TA image is composed as follows:
+```
+TA = (shdr | H | S | shdr_bootstrap_ta | standard ELF)
+```
+
+For OP-TEE 3.6.0, `shdr_encrypted_ta` does not exist and there only exists
+the Bootstrap TA image type.
+
+When OP-TEE reads a TA binary
+from the REE TA store, the store takes care of performing signature
+verification and decryption.
+
+By the time a TA binary reaches the TA loader proper (i.e., the component
+that lays out ELF sections onto memory), if the binary was loaded from
+the REE TA store, the TA binary header has already been stripped away
+and the entire binary is in cleartext. If the binary was loaded from
+either of the other two stores, the TA binary header will already never
+have been there to begin with.
+
+After the TA loader lays out the binary in memory and has had the TA
+instance initialize itself, a session handle is returned to the CA. This
+session handle is then used by the CA for subsequent communication with
+the TA instance.
+
+#### Image Validation
+
+When a TA is read from the REE's filesystem, the REE TA store checks that
+the signature of the hash is correct. Then, it computes the hash as per the
+formulas above, choosing which one depending on whether the TA is encrypted,
+and verifies that it matches the signed hash. If everything checks out, the
+cleartext, standard ELF binary is passed to the ELF loader proper.
+
+#### TA Stores & Open Enclave
+
+A TA is a TA regardless of where it is stored. The three TA stores, namely
+the REE TA store, the Secure Storage TA store, and the Early TA store,
+retrieve regular TA binaries. So, if you build a TA with the Open Enclave SDK,
+you can make that into an early TA by feeding it to OP-TEE's build system,
+which will embed the TA binary into OP-TEE itself. Similarly, you can take
+an Open Enclave TA and install it into secure storage.
```

Comment:
Do we have (or need) test cases that cover different types of stores? My understanding is that the existing test cases only use the REE TA store, which requires the TA header.

For supporting the Secure Storage TA store and the Early TA store, can the binary still come the TA header (i.e., use `OE_SET_ENCLAVE_OPTEE`)?

Reply:
We don't have test cases for neither the Secure Storage TA store nor for the Early TA store; we could add some. I admittedly have not tried these scenarios with the SDK.

For the Secure Storage TA store, the binary that you install as described in the next paragraph must have the TA header. For the Early TA store, I'll get back to you.

Reply:
For an Early TA store TA, the TA header is still there in the TA to be embedded into OP-TEE. OP-TEE's build system takes care of processing the TA binary and generating a C file with the binary as payload, which then gets linked into OP-TEE itself.